alpaca/hypothesis_test.go

```diff
@@ -0,0 +1,259 @@
+	})
+	req, _ := http.NewRequest(http.MethodGet, client.opts.BaseURL, nil)
+	// In defaultDo, OAuth header should be set:
+	req.Header.Set("Authorization", "Bearer "+client.opts.OAuth)
+	assert.Equal(t, "Bearer oauthtoken", req.Header.Get("Authorization"))
+
+	// If OAuth is not provided but Broker credentials are, then Basic Auth is used.
+	client = NewClient(ClientOpts{
+		BrokerKey:    "brokerkey",
+		BrokerSecret: "brokersecret",
+		BaseURL:      "http://dummy",
+	})
+	req, _ = http.NewRequest(http.MethodGet, client.opts.BaseURL, nil)
+	req.SetBasicAuth(client.opts.BrokerKey, client.opts.BrokerSecret)
+	authHeader := req.Header.Get("Authorization")
+	require.True(t, strings.HasPrefix(authHeader, "Basic "))
+	decoded, err := base64.URLEncoding.DecodeString(strings.TrimPrefix(authHeader, "Basic "))
+	require.NoError(t, err)
+	parts := strings.Split(string(decoded), ":")
+	assert.Equal(t, "brokerkey", parts[0])
+	assert.Equal(t, "brokersecret", parts[1])
+}
+
+// ---------------------------
+// Robust Retry & Error Injection
+// ---------------------------
+
+// Test the retry mechanism under sustained pressure (simulate intermittent server errors).
+func TestDefaultDo_RetryUnderPressure(t *testing.T) {
+	callCount := 0
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		callCount++
+		// Fail the first three calls, then succeed.
+		if callCount <= 3 {
+			http.Error(w, "Temporary Error", http.StatusTooManyRequests)
+			return
+		}
+		w.Write([]byte("final success"))
+	}))
+	defer ts.Close()
+
+	client := NewClient(ClientOpts{
+		BaseURL:    ts.URL,
+		RetryLimit: 5,
+		RetryDelay: 5 * time.Millisecond,
+	})
+
+	req, _ := http.NewRequest(http.MethodGet, ts.URL, nil)
+	resp, err := defaultDo(client, req)
+	require.NoError(t, err)
+	b, _ := io.ReadAll(resp.Body)
+	assert.Equal(t, "final success", string(b))
+	assert.GreaterOrEqual(t, callCount, 4)
+}
+
+// ---------------------------
+// Endpoint & Streaming Fuzzing
+// ---------------------------
+
+// Fuzz test for the streaming endpoint: provide random input lines and ensure the handler never crashes.
+func FuzzStreamTradeUpdates_StreamingParser(f *testing.F) {
+	// Seed with valid messages.
+	validMessages := []string{
+		`data: {"execution_id":"first","at":"2024-01-01T00:00:00Z"}` + "\n",
+		`data: {"execution_id":"second","at":"2024-01-01T00:00:01Z"}` + "\n",
+	}
+	for _, msg := range validMessages {
+		f.Add(msg)
+	}
+	f.Fuzz(func(t *testing.T, input string) {
+		// Wrap the input into a bufio.Reader to simulate stream reading.
+		reader := bufio.NewReader(strings.NewReader(input))
+		// Only process lines starting with "data: " to mimic our stream.
+		line, err := reader.ReadBytes('\n')
+		if err != nil {
+			return
+		}
+		if !bytes.HasPrefix(line, []byte("data: ")) {
+			return
+		}
+		line = line[len("data: "):]
+		var tu TradeUpdate
+		// The Unmarshal might error, but we should not panic.
+		_ = json.Unmarshal(line, &tu)
```

Comment:
We don't actually test anything here except that `json.Unmarshal` never panics, but that's not even part of this SDK.